bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/DeviceThingHandler.java

```diff
@@ -99,33 +98,33 @@ private void initializeGroupAddresses() {
     @Override
     public void dispose() {
         cancelChannelFutures();
-        freeGroupAdresses();
+        freeGroupAddresses();
         super.dispose();
     }
 
     private void cancelChannelFutures() {
-        for (ScheduledFuture<?> future : channelFutures.values()) {
-            if (!future.isDone()) {
-                future.cancel(true);
-            }
+        for (ChannelUID channelUID : channelFutures.keySet()) {
+            channelFutures.computeIfPresent(channelUID, (k, v) -> {
+                v.cancel(true);
+                return null;
+            });
         }
-        channelFutures.clear();
     }
 
-    private void freeGroupAdresses() {
+    private void freeGroupAddresses() {
         groupAddresses.clear();
         groupAddressesWriteBlockedOnce.clear();
         groupAddressesRespondingSpec.clear();
     }
 
     @Override
     protected void cancelReadFutures() {
-        for (ScheduledFuture<?> future : readFutures.values()) {
-            if (!future.isDone()) {
-                future.cancel(true);
-            }
+        for (GroupAddress groupAddress : readFutures.keySet()) {
+            readFutures.computeIfPresent(groupAddress, (k, v) -> {
+                v.cancel(true);
+                return null;
+            });
```

Comment:
Not sure if this really improves the readability of the code - as it would add another indirection. I might miss something, but I have seen this code block only twice. And cancelling Futures is only used in ```DeviceThingHandler.java``` and ```AbstractKNXClient.java```.